Add timeouts for builds

zaneduffield opened the pull request:
Fixes #322.

I mostly followed the suggested steps given in https://github.com/sourcefrog/cargo-mutants/issues/322#issuecomment-2050131753, with a couple of differences:
- I didn't find any existing unit tests for CLI options, so I didn't add any CLI unit tests for the new options
- I had to modify the rounding behaviour for timeout durations so that a very short build timeout wouldn't get rounded to zero (e.g. in the integration test of the `well_tested` tree)
- I set the default `build-timeout-multiplier` value to 2, rather than the suggested 5. I figure that since incremental builds are expected to be substantially faster than the initial build, we shouldn't expect to be waiting very long for a build.
- I didn't add a new `ScenarioOutcome` for a build timeout, mostly because in the default terminal output you can tell whether it was the build or the test that timed out by looking at the phase at the end of the line:
  - a build timeout looks like
    ```text
    TIMEOUT  src/lib.rs:12:5: replace should_stop_const -> bool with false in 1.0s build
    ``` 
  - a test timeout looks like
    ```text
    TIMEOUT  src/lib.rs:25:5: replace should_stop -> bool with false in 0.2s build + 1.0s test
    ``` 

In updating the `book/src` documentation, I felt like I would be repeating a lot of text if I were to create a new section just for build timeouts, so I rewrote the whole page to talk about both timeouts simultaneously.

zaneduffield commented:
I forgot to test after rebasing. I'll fix it later today.

zaneduffield commented:
The failing test passes on my system. I might need to implement a minimum for the build timeout to make the CI tests less flaky.

The reason I didn't do this at first was it made the `well_tested_tree_with_short_build_timeout` test difficult to implement. For that test, we really do want a short timeout that's less than any reasonable minimum.

I could perhaps copy the `--minimum-test-timeout` option (and potentially the associated environment variable) for the build timeout, but it didn't feel right to be adding new CLI options just to make testing easier.

sourcefrog commented:
> Fixes #322.
> 
> I mostly followed the suggested steps given in [#322 (comment)](https://github.com/sourcefrog/cargo-mutants/issues/322#issuecomment-2050131753), with a couple of differences:
> 
> * I didn't find any existing unit tests for CLI options, so I didn't add any CLI unit tests for the new options

There are some like this at the bottom of `options.rs` but since this is probably a pretty simple option covered by other things it may not really be necessary

https://github.com/sourcefrog/cargo-mutants/blob/9cf221652a74ac8dff8fbe95a01b5909a56b3e0d/src/options.rs#L297-L308

> * I had to modify the rounding behaviour for timeout durations so that a very short build timeout wouldn't get rounded to zero (e.g. in the integration test of the `well_tested` tree)
> * I set the default `build-timeout-multiplier` value to 2, rather than the suggested 5. I figure that since incremental builds are expected to be substantially faster than the initial build, we shouldn't expect to be waiting very long for a build.

Makes sense.

> * I didn't add a new `ScenarioOutcome` for a build timeout, mostly because in the default terminal output you can tell whether it was the build or the test that timed out by looking at the phase at the end of the line:
>   
>   * a build timeout looks like
>     ```
>     TIMEOUT  src/lib.rs:12:5: replace should_stop_const -> bool with false in 1.0s build
>     ```
>   * a test timeout looks like
>     ```
>     TIMEOUT  src/lib.rs:25:5: replace should_stop -> bool with false in 0.2s build + 1.0s test
>     ```

In general I want to make sure it's visible in structured json output too, but it probably is, by looking at which phase failed, so this seems fine. Anyhow, your motivating case of a const function shows that it's really fairly similar to a timeout in running the tests.

> 
> In updating the `book/src` documentation, I felt like I would be repeating a lot of text if I were to create a new section just for build timeouts, so I rewrote the whole page to talk about both timeouts simultaneously.

Sounds good.

sourcefrog commented on src/lab.rs:
```diff
@@ -190,33 +198,75 @@ pub fn test_mutants(
+fn build_timeout(baseline_duration: Option<Duration>, options: &Options) -> Duration {
+    phase_timeout(
+        options.build_timeout,
+        Duration::MAX,
```
This seems to say that with `--check`, i.e. running just `cargo check --tests` on each mutant, the build timeout is unlimited. I haven't checked yet but I suspect a const fn that hangs the compiler might also hang the checks? Maybe we should check that it doesn't.

And, maybe the logic for the check timeout should be the same as for non-check builds, i.e. a multiplier of the baseline?

zaneduffield replied:
My intent was for the `check` timeouts to always be the same as the `build`. It's definitely true that the `check` hangs when a const function is divergent.

I didn't really test that the `check` works like this, though. Might be an area to improve the testing in.

zaneduffield replied:
This code here was a little backwards, the code change here
https://github.com/sourcefrog/cargo-mutants/pull/345/files/9cf221652a74ac8dff8fbe95a01b5909a56b3e0d#diff-43aa7803b24379ea66e0395da5f5baf8ea6eef4bd298b984bfa7c992f0de312dR309
made sure that the timeout used for `check` phases is whatever the final `build` timeout is.

I'll delete the special handing here, and add some tests to assert how it works.

sourcefrog commented on src/lab.rs:
```diff
@@ -190,33 +198,75 @@ pub fn test_mutants(
-                .round() as u64,
-            ),
+        check_only_timeout
+    } else if let Some(baseline_duration) = baseline_duration {
```
I think somewhere in here we want to say, "if Options.in_place && phase == Build|Check" then don't use the baseline, give a warning and use the minimum.

That might be best done by giving this a phase enum parameter and folding the `build/test_timeout` functions into this, but see what works.

zaneduffield replied:
Good point, I didn't know that the baseline can be computed in-place.

sourcefrog commented:
Thanks, that looks good!

> tests for options

Oh I see you have some in `lab.rs`, that's fine.

I realized that one complication will be from `--in-place`: the tree may have been freshly build and so the baseline build will be very fast. 

One option there would be to treat it analogous to `--baseline=skip` and just assume a fairly generous default of say 300s, and print a warning.

sourcefrog commented:
Regarding the test failing with an unexpected timeout, <https://github.com/sourcefrog/cargo-mutants/actions/runs/8770143687/job/24066400701>

```

command=`env -u CARGO_MUTANTS_MINIMUM_TEST_TIMEOUT -u CARGO_TERM_COLOR "/home/runner/work/cargo-mutants/cargo-mutants/target/debug/cargo-mutants" "mutants" "--no-times" "--no-shuffle" "-d" "/tmp/.tmpRJLpMJ"`
code=3
stdout=```
Found 6 mutants to test
ok       Unmutated baseline
MISSED   src/b.rs:2:5: replace one_untested -> i32 with 0
MISSED   src/b.rs:2:5: replace one_untested -> i32 with -1
TIMEOUT  src/c.rs:2:5: replace one -> i32 with 0
6 mutants tested: 2 missed, 3 caught, 1 timeouts
```

I can't immediately see why that would timeout, but I guess it is just the build is fairly fast on the baseline and so the timeout is set low enough that it's slightly flaky. 

I think you could add `-Ltrace` to the arguments, so that we can see more in the stderr of the failure what was going on.

This might be something that could happen in user's CI on their own tree so I think we should change something to avoid it: maybe a minimum build timeout hardcoded to 60s?

These should be a lot less common that mutants causing hangs in runtime code so I think we could afford to set the minimum fairly high.

zaneduffield commented on src/options.rs:
```diff
@@ -212,6 +222,10 @@ impl Options {
+            build_timeout: args.build_timeout.map(Duration::from_secs_f64),
+            build_timeout_multiplier: config
+                .build_timeout_multiplier
+                .or(args.build_timeout_multiplier),
```
Sure, I'll change that in a subsequent commit.

zaneduffield commented on src/lab.rs:
```diff
@@ -322,6 +384,22 @@ mod test {
+            Duration::from_secs(42 * 2),
+        );
+    }
+
```
I think that would be fine as a unit test in `options.rs`.

zaneduffield commented on tests/main.rs:
```diff
@@ -218,6 +221,28 @@ fn well_tested_tree_finds_no_problems() {
 }
 
+#[test]
+fn well_tested_tree_with_short_build_timeout() {
```
Good point, it probably is adequately exercises elsewhere. Removing this test would make it easier to make the tests less flaky, so I'll go for that approach.

zaneduffield commented on tests/main.rs:
```diff
@@ -775,7 +804,7 @@ fn mutants_causing_tests_to_hang_are_stopped_by_manual_timeout() {
             .expect("parse outcomes.json");
-    assert_eq!(outcomes_json["timeout"], 1);
+    assert_eq!(outcomes_json["timeout"], 2);
 }
```
It was a bit fiddly, but I got it working.

zaneduffield commented on src/lab.rs:
```diff
@@ -190,33 +198,75 @@ pub fn test_mutants(
+            Duration::from_secs_f64(baseline_duration.as_secs_f64() * multiplier),
         );
+
+        // Round large timeouts to a nice even number, but round smaller timeouts to 2DP
```
Will remove.

zaneduffield commented on src/lab.rs:
```diff
@@ -190,33 +198,75 @@ pub fn test_mutants(
+
+        // Round large timeouts to a nice even number, but round smaller timeouts to 2DP
+        timeout = if timeout.as_secs() >= 1 {
+            Duration::from_secs(timeout.as_secs_f64().round() as u64)
```
`ceil` would be more appropriate here! That would have been a much simpler way to do it. I'll replace the original code, with `round` substituted for `ceil`.

zaneduffield commented on src/main.rs:
```diff
@@ -290,6 +290,14 @@ pub struct Args {
+    build_timeout: Option<f64>,
+
+    /// build timeout multiplier (relative to base build time).
+    #[arg(long, help_heading = "Execution")]
```
Sure, I'll also put that in a subsequent commit.

zaneduffield commented on src/lab.rs:
```diff
@@ -344,4 +434,15 @@ mod test {
+        let options = Options::new(&args, &Config::default()).unwrap();
+
+        assert_eq!(options.build_timeout_multiplier, None);
+        assert_eq!(build_timeout(None, &options), Duration::from_secs(300),);
```
The 300s fallback is used when
- the `explicit_timeout` (i.e. `--timeout`/`--build-timeout`) is None, and
- `check_only` is false, and
- there is no `baseline_duration`

I was just following the original code here. I feel like the message printed when this 300s value is chosen is pretty clear.

I'll add a line about this in the book if that helps.

zaneduffield commented:
I haven't finished responding to your review, but I've done as much as I can for today and pushed it. I'll return later this week and try and finish it.

zaneduffield commented:
The test failure now is a bit weird...
https://github.com/sourcefrog/cargo-mutants/actions/runs/8799621313/job/24149040400?pr=345#step:9:149
```
---- output::test::create_output_dir stdout ----
thread 'output::test::create_output_dir' panicked at src/output.rs:275:51:
called `Result::unwrap()` on an `Err` value: run cargo locate-project in "/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.tmpqosAJJ"
error: test failed, to rerun pass `-p cargo-mutants --bin cargo-mutants`

Caused by:
    ["/Users/runner/.rustup/toolchains/1.74-x86_64-apple-darwin/bin/cargo", "locate-project", "--workspace"] failed with status Signaled(4): 
```

I'm not really sure what that's about, but I think I've addressed all the review comments now.

sourcefrog commented:
cargo seems to be failing with signal 4, SIGILL, "illegal instruction". I see this sometimes on the GitHub actions and I don't know a cause for it yet, but it seems to be intermittent and fixed by a rebuild. Maybe it's some issue in their mac virtualization (if any).

With that fixed it looks like there is one missed mutant

`MISSED   src/lab.rs:225:46: replace != with == in phase_timeout in 6.8s build + 93.0s test`

Please see if you can add a reasonable test for it; if not just let me know.

zaneduffield commented:
> With that fixed it looks like there is one missed mutant
> 
> `MISSED src/lab.rs:225:46: replace != with == in phase_timeout in 6.8s build + 93.0s test`
> 
> Please see if you can add a reasonable test for it; if not just let me know.

Done

zaneduffield commented:
Also, this PR fixes #323.

sourcefrog commented:
Thanks, I'll look on the weekend if not before.
